037-pluggable-pod-security-profiles.md

```diff
@@ -0,0 +1,234 @@
+The methods providing the security context will get as a parameter an object of type `PodSecurityProviderContext`.
+This will be an object created by Strimzi which will encapsulate the different infromation which strimzi will provide to the Provider plugin to make its decisions when generating the Kubernetes security context.
+The `PodSecurityProviderContext` will be defines as another interface in the `api` module:
+
+```java
+interface PodSecurityProviderContext {
+    Storage storage();
+    PodSecurityContext userSuppliedContext();
+}
+```
+
+This encapsulation should make it easier to add additional information in the future without changing the signature of the provider methods but only by adding new methods to the Strimzi provided context.
+
+The provider methods will have the default implementation which will always just return the user-specified context.
+For example (just few methods are listed in this example):
+
+```java
+public interface PodSecurityProvider {
+    // ...
+
+    default PodSecurityContext kafkaPodSecurityContext(PodSecurityProviderContext context) {
+        return userSecurityContext;
+    }
+
+    default SecurityContext kafkaContainerSecurityContext(PodSecurityProviderContext context) {
+        return userSecurityContext;
+    }
+
+    default SecurityContext kafkaInitContainerSecurityContext(PodSecurityProviderContext context) {
+        return userSecurityContext;
+    }
+
+    // ...
+}
+```
+
+The interface will also have a `configure` method for configuring the provider.
+The `configure` method will get as a parameter an object describing the platform (Kubernetes version, features / APIs) which can be taken into account when creating the security context.
+This method will not have any default implementation.
+
+```java
+public interface PodSecurityProvider {
+    // ...
+    void configure(PlatformFeatures platformFeatures);
+    // ...
+}
+```
+
+`PlatformFeatures` in this method is another new interface which provides some of the methods provided by the `PlatformFeaturesAvailability` class which is part of the `operator-common` module`
+It uses the features which might be relevant for generating the securitny context:
+
+```java
+public interface PlatformFeatures {
+    KubernetesVersion getKubernetesVersion();
+    boolean isOpenshift();
+}
+```
+
+The `configure(...)` method can in addition to the `PlatformFeatures` also for example use environment variables to configure the provider.
+
+#### Implementations
+
+We will provide two implementations to support the baseline (default) and restricted profiles:
+* `BaselinePodSecurityProvider`
+* `RestrictedPodSecurityProvider`
+
+These implementations will be also part of the `api` module so that users can just extend them instead of implementing the provider from scratch.
+
+The `BaselinePodSecurityProvider` will implement the current behaviour:
+* Return the user-configured security context when specified
```

Comment:
I think `userSuppliedContext()` in the above snippets and `user-configured` in this section are the same concept, so we should be consistent with naming.